diff -up util-linux-ng-2.17.2/mount/umount.c.kzak util-linux-ng-2.17.2/mount/umount.c
--- util-linux-ng-2.17.2/mount/umount.c.kzak	2015-12-17 14:29:09.101375133 +0100
+++ util-linux-ng-2.17.2/mount/umount.c	2015-12-17 14:30:03.761983076 +0100
@@ -518,11 +518,15 @@ umount_file (char *arg) {
 		return 0;
 	}
 
-	file = canonicalize(arg); /* mtab paths are canonicalized */
 	if (verbose > 1)
-		printf(_("Trying to umount %s\n"), file);
+		printf(_("Trying to umount %s\n"), arg);
 
-	mc = getmntdirbackward(file, NULL);
+	mc = getmntdirbackward(arg, NULL);
+	if (!mc) {
+		file = canonicalize(arg);
+		mc = getmntdirbackward(file, NULL);
+	} else
+		file = arg;
 	if (!mc) {
 		mc = getmntdevbackward(file, NULL);
 		if (mc) {
