diff -up util-linux-ng-2.17.2/misc-utils/blkid.c.kzak util-linux-ng-2.17.2/misc-utils/blkid.c
--- util-linux-ng-2.17.2/misc-utils/blkid.c.kzak	2011-08-10 15:58:54.012695629 +0200
+++ util-linux-ng-2.17.2/misc-utils/blkid.c	2011-08-10 15:59:17.899999668 +0200
@@ -578,7 +578,7 @@ static int list_to_usage(const char *lis
 int main(int argc, char **argv)
 {
 	blkid_cache cache = NULL;
-	char *devices[128] = { NULL, };
+	char **devices = NULL;
 	char *show[128] = { NULL, };
 	char *search_type = NULL, *search_value = NULL;
 	char *read = NULL;
@@ -693,6 +693,16 @@ int main(int argc, char **argv)
 			usage(err);
 		}
 
+
+	/* The rest of the args are device names */
+	if (optind < argc) {
+		devices = calloc(argc - optind, sizeof(char *));
+		if (!devices) {
+			fprintf(stderr, "Failed to allocate device name array\n");
+			goto exit;
+		}
+	}
+
 	while (optind < argc)
 		devices[numdev++] = argv[optind++];
 
@@ -828,5 +838,6 @@ exit:
 	free(search_value);
 	if (!lowprobe && !eval)
 		blkid_put_cache(cache);
+	free(devices);
 	return err;
 }
