diff -up util-linux-ng-2.17.2/mount/mount.c.kzak util-linux-ng-2.17.2/mount/mount.c
--- util-linux-ng-2.17.2/mount/mount.c.kzak	2013-07-15 11:54:22.362665402 +0200
+++ util-linux-ng-2.17.2/mount/mount.c	2013-07-15 11:55:08.892063550 +0200
@@ -1969,11 +1969,12 @@ do_mount_all (char *types, char *options
 		/* if child, or not forked, do the mounting */
 		if (p == 0 || p == -1) {
 			for (mc = cp->mec; mc; mc = mc->nxt) {
-				status |= mount_one (mc->m.mnt_fsname,
+				int res = mount_one (mc->m.mnt_fsname,
 						     mc->m.mnt_dir,
 						     mc->m.mnt_type,
 						     mc->m.mnt_opts,
 						     options, 0, 0);
+				status |= res == 0 ? EX_SOMEOK : res;
 			}
 			if (mountcount)
 				status |= EX_SOMEOK;
