diff -up util-linux-ng-2.17.2/mount/lomount.c.kzak util-linux-ng-2.17.2/mount/lomount.c
--- util-linux-ng-2.17.2/mount/lomount.c.kzak	2010-03-18 23:11:23.000000000 +0100
+++ util-linux-ng-2.17.2/mount/lomount.c	2011-01-12 11:50:40.000000000 +0100
@@ -882,8 +882,8 @@ find_unused_loop_device (void) {
 #include <stdarg.h>
 
 static void
-usage(void) {
-	fprintf(stderr, _("\nUsage:\n"
+usage(FILE *f) {
+	fprintf(f, _("\nUsage:\n"
   " %1$s loop_device                             give info\n"
   " %1$s -a | --all                              list all used\n"
   " %1$s -d | --detach <loopdev> [<loopdev> ...] delete\n"
@@ -893,7 +893,7 @@ usage(void) {
   " %1$s [ options ] {-f|--find|loopdev} <file>  setup\n"),
 		progname);
 
-	fprintf(stderr, _("\nOptions:\n"
+	fprintf(f, _("\nOptions:\n"
   " -e | --encryption <type> enable data encryption with specified <name/num>\n"
   " -h | --help              this help\n"
   " -o | --offset <num>      start at offset <num> into file\n"
@@ -902,7 +902,8 @@ usage(void) {
   " -r | --read-only         setup read-only loop device\n"
   "      --show              print device name (with -f <file>)\n"
   " -v | --verbose           verbose mode\n\n"));
-	exit(1);
+
+	exit(f == stderr ? EXIT_FAILURE : EXIT_SUCCESS);
  }
 
 int
@@ -966,6 +967,9 @@ main(int argc, char **argv) {
 		case 'f':
 			find = 1;
 			break;
+		case 'h':
+			usage(stdout);
+			break;
 		case 'j':
 			assoc = optarg;
 			break;
@@ -987,39 +991,39 @@ main(int argc, char **argv) {
                         break;
 
 		default:
-			usage();
+			usage(stderr);
 		}
 	}
 
 	if (argc == 1) {
-		usage();
+		usage(stderr);
 	} else if (delete) {
 		if (argc < optind+1 || encryption || offset || sizelimit ||
 		    capacity || find || all || showdev || assoc || ro)
-			usage();
+			usage(stderr);
 	} else if (find) {
 		if (capacity || all || assoc || argc < optind || argc > optind+1)
-			usage();
+			usage(stderr);
 	} else if (all) {
 		if (argc > 2)
-			usage();
+			usage(stderr);
 	} else if (assoc) {
 		if (capacity || encryption || showdev || passfd || ro)
-			usage();
+			usage(stderr);
 	} else if (capacity) {
 		if (argc != optind + 1 || encryption || offset || sizelimit ||
 		    showdev || ro)
-			usage();
+			usage(stderr);
 	} else {
 		if (argc < optind+1 || argc > optind+2)
-			usage();
+			usage(stderr);
 	}
 
 	if (offset && sscanf(offset, "%llu", &off) != 1)
-		usage();
+		usage(stderr);
 
 	if (sizelimit && sscanf(sizelimit, "%llu", &slimit) != 1)
-		usage();
+		usage(stderr);
 
 	if (all)
 		return show_used_loop_devices();
@@ -1053,7 +1057,7 @@ main(int argc, char **argv) {
 		res = show_loop(device);
 	else {
 		if (passfd && sscanf(passfd, "%d", &pfd) != 1)
-			usage();
+			usage(stderr);
 		do {
 			res = set_loop(device, file, off, slimit, encryption, pfd, &ro);
 			if (res == 2 && find) {
