diff -up util-linux-ng-2.17.2/login-utils/agetty.8.kzak util-linux-ng-2.17.2/login-utils/agetty.8
--- util-linux-ng-2.17.2/login-utils/agetty.8.kzak	2011-08-11 11:50:26.989811074 +0200
+++ util-linux-ng-2.17.2/login-utils/agetty.8	2011-08-11 11:51:29.628585248 +0200
@@ -106,6 +106,10 @@ login prompt. Terminals or communication
 when receiving lots of text at the wrong baud rate; dial-up scripts
 may fail if the login prompt is preceded by too much text.
 .TP
+\-E
+If \-H \fIfakehost\fP option is given then \-h \fIfakehost\fP options is
+added to the the \fB/bin/login\fP command line.
+.TP
 \-f \fIissue_file\fP
 Display the contents of \fIissue_file\fP instead of \fI/etc/issue\fP.
 This allows custom messages to be displayed on different terminals.
diff -up util-linux-ng-2.17.2/login-utils/agetty.c.kzak util-linux-ng-2.17.2/login-utils/agetty.c
--- util-linux-ng-2.17.2/login-utils/agetty.c.kzak	2011-08-11 11:44:36.501434276 +0200
+++ util-linux-ng-2.17.2/login-utils/agetty.c	2011-08-11 11:53:40.801189111 +0200
@@ -134,6 +134,11 @@ struct options {
 #define F_NOPROMPT	(1<<7)		/* don't ask for login name! */
 #define F_LCUC		(1<<8)		/* Support for *LCUC stty modes */
 
+/*
+ * backport from util-linux 2.21 to RHEL6
+ */
+#define F_REMOTE	(1<<20)		/* Add '-h fakehost' to login(1) command line */
+
 /* Storage for things detected while the login name was read. */
 
 struct chardata {
@@ -356,7 +361,11 @@ main(argc, argv)
 
     /* Let the login program take care of password validation. */
 
-    (void) execl(options.login, options.login, "--", logname, NULL);
+    if ((options.flags & F_REMOTE) && fakehost)
+        execl(options.login, options.login, "-h", fakehost, "--", logname, NULL);
+    else
+        execl(options.login, options.login, "--", logname, NULL);
+
     error(_("%s: can't exec %s: %m"), options.tty, options.login);
     exit(0);  /* quiet GCC */
 }
@@ -373,7 +382,7 @@ parse_args(argc, argv, op)
     extern int optind;			/* getopt */
     int     c;
 
-    while (isascii(c = getopt(argc, argv, "8I:LH:f:hil:mt:wUn"))) {
+    while (isascii(c = getopt(argc, argv, "8I:LH:Ef:hil:mt:wUn"))) {
 	switch (c) {
 	case '8':
 	    op->eightbits = 1;
@@ -424,6 +433,9 @@ parse_args(argc, argv, op)
 	case 'H':                               /* fake login host */
 	    fakehost = optarg;
 	    break;
+	case 'E':
+	    op->flags |= F_REMOTE;
+	    break;
 	case 'f':				/* custom issue file */
 	    op->flags |= F_CUSTISSUE;
 	    op->issue = optarg;
